### PR TITLE
Passphrase Redesign

### PR DESCRIPTION
**Pull request by tsusanka**
Replaces #714.

This is still not intended to be merged into master at the moment. @szymonlesisz is currently working on this in Connect, so let's see if he has some more comments on this first.

In the meantime it would be great to get some review on this.

### CODE REVIEW

File: common/protob/messages-common.proto [hunk 1]
```diff
@@ -110,31 +111,16 @@ message PinMatrixAck {
  * @next PassphraseAck
  */
 message PassphraseRequest {
-    optional bool on_device = 1;    // passphrase is being entered on the device
+    optional bool _deprecated_on_device = 1;  // moved to PassphraseAck but left for backwards compatibility
```
Comment (matejcik) [inline common/protob/messages-common.proto]: didn't we decide in the end that this will be left out?

Comment (tsusanka) [reply common/protob/messages-common.proto]: I wasn't sure to be honest. I wanted to double-check with you

File: core/src/apps/common/passphrase.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    assert isinstance(passphrase, str)
+
+    return passphrase
+
+
+async def _entry_dialog() -> None:
+    text = Text("Passphrase entry", ICON_CONFIG)
+    text.normal("Please type your", "passphrase on the", "connected host.")
+    # no need to specify timeout, because it hangs till PassphraseAck is received
+    # TODO ask: not sure this is correct. If we change the behaviour of Popup
```
Comment (tsusanka) [inline core/src/apps/common/passphrase.py]: cc @matejcik

File: legacy/firmware/protect.c [hunk 3]
```diff
@@ -369,12 +370,25 @@ bool protectPassphrase(void) {
...
+      }
+      if (!ppa->has_passphrase) {
+        fsm_sendFailure(FailureType_Failure_DataError,
+                        _("No passphrase provided. Use empty string to set an "
+                          "empty passphrase."));
+        result = false;
+        break;
+      }
+      // TODO: ask - why ppa->passphrase.size is not working? because of tiny?
+      strlcpy(passphrase, ppa->passphrase, sizeof(ppa->passphrase));
```
Comment (tsusanka) [inline legacy/firmware/protect.c]: cc @prusnak

File: common/protob/messages-common.proto (hunk 1 above)
Comment (matejcik) [reply common/protob/messages-common.proto]: I originally wanted to keep this for backwards compatibility, BUT removing `PassphraseState` messages breaks that irreparably anyway
so if we wanted backwards compatibility that bad, we would need to keep these too -- which we don't want, I think?

File: tests/device_tests/test_msg_clearsession.py [hunk 4]
```diff
@@ -17,77 +17,64 @@
...
-                "random data",
-                button_protection=True,
-                pin_protection=True,
-                passphrase_protection=True,
-            )
-            assert res == "random data"
+@pytest.mark.skip_ui
+@pytest.mark.setup_client(pin=True, passphrase=True)
+def test_clear_session(client):
+    if client.features.model == "1":
```
Comment (tsusanka) [inline tests/device_tests/test_msg_clearsession.py]: @matejcik Maybe there is a nicer way?

File: common/protob/messages-common.proto (hunk 1 above)
Comment (tsusanka) [reply common/protob/messages-common.proto]: Oh, yeah! Yes, I think it's time to make break it. I will remove this.

File: core/src/apps/homescreen/__init__.py [hunk 5]
```diff
@@ -67,18 +67,19 @@ def get_features() -> Features:
...
     f.sd_protection = storage.sd_salt.is_enabled()
     f.wipe_code_protection = config.has_wipe_code()
+    f.session_id = cache.get_session_id()
+    f.passphrase_always_on_device = storage.device.get_passphrase_always_on_device()
     return f
 
 
 async def handle_Initialize(ctx: wire.Context, msg: Initialize) -> Features:
-    if msg.state is None or msg.state != cache.get_state(prev_state=bytes(msg.state)):
+    if msg.session_id is None or msg.session_id != cache.get_session_id():
```
Comment (matejcik) [inline core/src/apps/homescreen/__init__.py]: if the client sends session id, and it's different from the current session, then they get a new random session

I'm pretty sure that this is what we want, but wanted to ask anyway: are there reasons for/against using the client-provided session id?

Comment (matejcik) [reply core/src/apps/homescreen/__init__.py]: one comes to mind: host can tell whether or not their session was active on the device

File: common/protob/messages-common.proto [hunk 6]
```diff
@@ -62,11 +62,12 @@ message ButtonRequest {
         ButtonRequest_PublicKey = 11;
         ButtonRequest_MnemonicWordCount = 12;
         ButtonRequest_MnemonicInput = 13;
-        ButtonRequest_PassphraseType = 14;
+//        ButtonRequest_PassphraseType = 14;  DEPRECATED
```
Comment (tsusanka) [inline common/protob/messages-common.proto]: Is this the right way or do we just remove it completely?

File: core/src/apps/management/apply_settings.py [hunk 7]
```diff
@@ -69,16 +71,16 @@
...
-        desc = "ASK"
+async def require_confirm_change_passphrase_source(
+    ctx, passphrase_always_on_device: bool
+):
     text = Text("Passphrase source", ui.ICON_CONFIG)
-    text.normal("Do you really want to", "change the passphrase", "source to")
-    text.bold("ALWAYS %s?" % desc)
+    if passphrase_always_on_device:
+        text.normal(
+            "Do you really want to", "entry passphrase always", "on the device?"
```
Comment (matejcik) [inline core/src/apps/management/apply_settings.py]: typo: "enter"

but the wording is not great either. let's have someone from a more wording-y department look this over?

File: common/protob/messages-common.proto (hunk 6 above)
Comment (matejcik) [reply common/protob/messages-common.proto]: i'd keep it like this

this whole thing will go away soon anyway .)

File: core/src/apps/monero/live_refresh.py [hunk 8]
```diff
@@ -49,10 +49,10 @@ def __init__(self):
         ctx, misc.validate_full_path, keychain, msg.address_n, CURVE
     )
 
-    passphrase_fprint = get_passphrase_fprint()
-    if live_refresh_token() != passphrase_fprint:
+    fingerprint = storage.cache.get_session_id()[:4]
```
Comment (matejcik) [inline core/src/apps/monero/live_refresh.py]: this is doing something very different from the original code
namely: in the original code, you could get `passphrase_fprint` by knowing the passphrase + device_id; i.e., it was possible to live-refresh on a cold-started device

now you need to know the current session_id, but not the passphrase

i didn't look into Monero code deep enough to know what this means, so making a note here

File: tests/device_tests/test_protection_levels.py [hunk 9]
```diff
@@ -63,18 +63,11 @@ def test_change_pin(self, client):
             )
             device.change_pin(client)
 
-    @pytest.mark.setup_client(pin=True, passphrase=True)
+    @pytest.mark.setup_client()
```
Comment (matejcik) [inline tests/device_tests/test_protection_levels.py]: empty setup_client is never needed

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+# You should have received a copy of the License along with this library.
+# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+
+import pytest
+
+from trezorlib import messages
+from trezorlib.tools import parse_path
+
+
+def _get_xpub(client, passphrase):
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: there should be an explanatory comment saying that all the calls are raw here (as opposed to using library functions) to avoid trezorlib's implicit session management (clearing session after settings, hidden handling of session ids, etc)

also there should not be `setup_client` decorators anywhere, as empty setup_client should never be needed

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    _enable_passphrase(client)
+
+    # The session id is unchanged, therefore we do not prompt for the passphrase.
+    response = client.call_raw(messages.Initialize(session_id=session_id))
+    xpub = _get_xpub(client, passphrase=None)
+    assert isinstance(response, messages.Features)
+    assert session_id == response.session_id
+    assert (
+        xpub
+        == "xpub6BiVtCpG9fQPxnPmHXG8PhtzQdWC2Su4qWu6XW9tpWFYhxydCLJGrWBJZ5H6qTAHdPQ7pQhtpjiYZVZARo14qHiay2fvrX996oEP42u8wZy"
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: please use constants declared at start for the xpubs.
for instance, here it is not clear at first look that this is the same xpub as above

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    assert isinstance(response, messages.Features)
+
+    # Turn on passphrase.
+    _enable_passphrase(client)
+
+    # This setting makes no sense on T1.
+    response = client.call_raw(messages.ApplySettings(passphrase_always_on_device=True))
+
+    assert isinstance(response, messages.Failure)
+    assert response.code == 3  # DataError
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: maybe use `FailureType.DataError` instead of 3?

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+
+@pytest.mark.skip_ui
+@pytest.mark.setup_client(passphrase=True)
+def test_passphrase_missing(client):
+    response = client.call_raw(
+        messages.GetPublicKey(address_n=parse_path("44'/0'/0'"), coin_name="Bitcoin")
+    )
+    assert isinstance(response, messages.PassphraseRequest)
+    response = client.call_raw(messages.PassphraseAck(passphrase=None))
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: maybe just `PassphraseAck()` ?

maybe also add a test case for `PassphraseAck(passphrase=None, on_device=False)` ?

File: core/src/apps/common/passphrase.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        passphrase = await ctx.wait(keyboard)
+    if passphrase is CANCELLED:
+        raise wire.ActionCancelled("Passphrase entry cancelled")
+
+    assert isinstance(passphrase, str)
+
+    return passphrase
+
+
+async def _entry_dialog() -> None:
```
Comment (matejcik) [inline core/src/apps/common/passphrase.py]: let's drop the `async`, return the `Popup` object, and rename the function to `_entry_popup`. That way the call site will be less confusing.

File: core/src/apps/common/passphrase.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        raise wire.DataError("Maximum passphrase length is %d" % _MAX_PASSPHRASE_LEN)
+
+    return passphrase
+
+
+async def _get_from_user(ctx: wire.Context) -> str:
+    if storage.device.get_passphrase_always_on_device():
+        return await request_from_user_on_device(ctx)
+
+    await _entry_dialog()
```
Comment (matejcik) [inline core/src/apps/common/passphrase.py]: starting at this line, until the end, this should be a separate function `_request_from_host` or something like that

File: core/src/apps/common/passphrase.py (hunk 14 above)
Comment (matejcik) [reply core/src/apps/common/passphrase.py]: unless we decide to do it differently (see below)

File: core/src/apps/common/passphrase.py (hunk 2 above)
Comment (matejcik) [reply core/src/apps/common/passphrase.py]: well, this is certainly interesting.

ISTM this should just draw on the screen, without using the layout mechanism -- like mnemonic.py does it at https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/common/mnemonic.py#L57
and then call `text.on_render()` directly

or better, put a bunch of utility methods in, uh, `ui/__init__` presumably? that do this.

basically what `handle_rendering()` does but without the busy loop?

File: core/src/apps/homescreen/__init__.py (hunk 5 above)
Comment (tsusanka) [reply core/src/apps/homescreen/__init__.py]: You mean to return the same session id as provided by the host even though it was not cached?

That is problematic, consider this behaviour:

- Host sends Initialize(), receives Features(session_id=A) with session id A.
- Someone else sends Initialize(), receives Features(session_id=B) and stuff happens.
- Host sends Initialize(session_id=A) which they stored and would receive Features(session_id=A), but the cache was cleared and Trezor needs to ask for passphrase again, which is wrong! The Host assumes that if the session id from Trezor is the same as the one they sent that it is the same session.

Does it make sense?

File: core/src/apps/management/apply_settings.py (hunk 7 above)
Comment (tsusanka) [reply core/src/apps/management/apply_settings.py]: Yes. I've mentioned this to @matejzak, he made a note that he will address it when they do copy in Suite.

File: core/src/apps/monero/live_refresh.py (hunk 8 above)
Comment (tsusanka) [reply core/src/apps/monero/live_refresh.py]: The session_id is always coupled to a single passphrase. I've talked to @ph4r05 before briefly and he mentioned he needs just a simple identifier of the session and he seemed fine with this solution. But let's cc him on this to make double sure.

File: tests/device_tests/test_session_id_and_passphrase.py (hunk 13 above)
Comment (tsusanka) [reply tests/device_tests/test_session_id_and_passphrase.py]: > maybe just PassphraseAck() ?

I kind of like it that it is explicit.

> maybe also add a test case for PassphraseAck(passphrase=None, on_device=False) ?

Will add.

File: core/src/storage/device.py [hunk 16]
```diff
@@ -143,21 +143,24 @@ def no_backup() -> bool:
...
 
 
-def get_passphrase_source() -> int:
-    b = common.get(_NAMESPACE, _PASSPHRASE_SOURCE)
-    if b == b"\x01":
-        return 1
-    elif b == b"\x02":
-        return 2
+def get_passphrase_always_on_device() -> bool:
+    b = common.get(_NAMESPACE, _PASSPHRASE_ALWAYS_ON_DEVICE)
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/storage/device.py]: Done, please check.

File: core/src/apps/homescreen/__init__.py (hunk 5 above)
Comment (matejcik) [reply core/src/apps/homescreen/__init__.py]: well, my question was mostly "do we actually want this behavior?"
what you described is a usecase where we do want it. so i suppose that's a yes

File: core/src/apps/monero/live_refresh.py (hunk 8 above)
Comment (ph4r05) [reply core/src/apps/monero/live_refresh.py]: @tsusanka correct, I just need some session ID (which changes per `(Trezor boot, passphrase)`) so I don't have to annoy users by asking for confirmation every minute.

Comment (matejcik) [reply core/src/apps/monero/live_refresh.py]: interesting
this highlights a usecase for something like `storage.cache.set(key, value)` that would bind the key-value pair to the session, and would go away with ClearSession or with passphrase change.

as it happens Cardano's `_cached_root` could also use that, and I'm afraid the current implementation is either incorrect, or only works through magic (see other comment)

File: core/src/apps/cardano/seed.py [hunk 17]
```diff
@@ -1,11 +1,15 @@
...
 
 from apps.cardano import CURVE, SEED_NAMESPACE
 from apps.common import mnemonic
-from apps.common.request_passphrase import protect_by_passphrase
+from apps.common.passphrase import get as get_passphrase
+
+if False:
+    from typing import Optional
+
+_cached_root = None  # type: Optional[bytes]
```
Comment (matejcik) [inline core/src/apps/cardano/seed.py]: as I just noticed: this variable lives exactly as long as `apps.cardano.seed` is imported

that means that either (a) the unimport mechanism is working and `apps.cardano.seed` goes away after every workflow - meaning that Cardano users would need to enter passphrase before each and every operation

or (b) it survives, which would be a bug in itself -- but then doesn't respect session management and stays the same over session and passphrase changes

File: core/src/storage/cache.py [hunk 18]
```diff
@@ -1,75 +1,41 @@
...
 
 if False:
     from typing import Optional
 
 _cached_seed = None  # type: Optional[bytes]
-_cached_seed_without_passphrase = None  # type: Optional[bytes]
-_cached_passphrase = None  # type: Optional[str]
-_cached_passphrase_fprint = b"\x00\x00\x00\x00"  # type: bytes
+_cached_seed_without_passphrase = None  # type: Optional[bytes]  # Needed for SLIP-21
+_cached_session_id = None  # type: Optional[bytes]
```
Comment (matejcik) [inline core/src/storage/cache.py]: would it make sense to convert some of these to elements in a dict? and have `storage.cache.get`, `set`, `delete`, sort of like `storage.common` does? this would get thrown away when session id changes.

perhaps similarly to the general storage, this should be structured based on app+key?

`session_id` should probably stay separate in any case

File: core/src/apps/homescreen/__init__.py (hunk 5 above)
Comment (tsusanka) [reply core/src/apps/homescreen/__init__.py]: Yes :)

File: core/src/apps/cardano/seed.py (hunk 17 above)
Comment (tsusanka) [reply core/src/apps/cardano/seed.py]: It seems (a) is correct.

I like the idea you have suggested above (`storage.cache.set(key, value)`), I will implement that now.

File: core/src/storage/cache.py (hunk 18 above)
Comment (tsusanka) [reply core/src/storage/cache.py]: Check out https://github.com/trezor/trezor-firmware/pull/803/commits/d112c2d266d9d0ffbb6155bd6e7610c199be16d4

File: core/src/apps/cardano/seed.py (hunk 17 above)
Comment (tsusanka) [reply core/src/apps/cardano/seed.py]: Check out https://github.com/trezor/trezor-firmware/pull/803/commits/d112c2d266d9d0ffbb6155bd6e7610c199be16d4

File: core/src/apps/cardano/seed.py [hunk 19]
```diff
@@ -30,30 +30,27 @@ def derive(self, node_path: list) -> bip32.HDNode:
...
+        else:
             seed = mnemonic.get_seed(passphrase)
-            storage.cache.set_seed(seed)
-        root = bip32.from_seed(seed, "ed25519 cardano seed")
+            root = bip32.from_seed(seed, "ed25519 cardano seed")
+
+        storage.cache.set(cache.APP_CARDANO_ROOT, root)
+
+    # let's not modify the one in the cache
+    root = root.clone()
```
Comment (matejcik) [inline core/src/apps/cardano/seed.py]: what's this about btw? couldn't we cache the derived root instead?

File: core/src/apps/common/passphrase.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+
+async def get(ctx: wire.Context) -> str:
+    if is_enabled():
+        return await _request_from_user(ctx)
+    else:
+        return ""
+
+
+async def _request_from_user(ctx: wire.Context) -> str:
```
Comment (matejcik) [inline core/src/apps/common/passphrase.py]: i still think `_request_from_user` and `_get_from_user` should be a single function: first get from user or from host, then check the length

File: core/src/apps/cardano/seed.py (hunk 19 above)
Comment (tsusanka) [reply core/src/apps/cardano/seed.py]: Indeed we should. I am not sure why I have solved it this way, we should cache the derived(SEED_NAMESPACE) instead. Fixed in bc1fd06abab4ad92bde84249406326d38dfe937e

File: core/src/apps/common/passphrase.py (hunk 20 above)
Comment (tsusanka) [reply core/src/apps/common/passphrase.py]: Definitely. Fixed in 61e6d3e4843f2a9000956b4e46225868b1da3c3e

File: core/src/apps/homescreen/__init__.py [hunk 21]
```diff
@@ -91,7 +92,7 @@ def get_features() -> Features:
 
 
 async def handle_ClearSession(ctx: wire.Context, msg: ClearSession) -> Success:
-    cache.clear(keep_passphrase=True)
```
Comment (matejcik) [inline core/src/apps/homescreen/__init__.py]: why did ClearSession keep the passphrase in the past?

Comment (tsusanka) [reply core/src/apps/homescreen/__init__.py]: Let's ask @prusnak if he remembers, I don't recall.

File: legacy/firmware/config.c [hunk 22]
```diff
@@ -606,58 +609,17 @@ bool config_getU2FRoot(HDNode *node) {
   return ret;
 }
 
-bool config_getRootNode(HDNode *node, const char *curve, bool usePassphrase) {
-  // if storage has node, decrypt and use it
-  StorageHDNode storageHDNode = {0};
```
Comment (matejcik) [inline legacy/firmware/config.c]: is this going away because XPRV support is gone?

File: tests/device_tests/test_msg_clearsession.py (hunk 4 above)
Comment (matejcik) [reply tests/device_tests/test_msg_clearsession.py]: i don't think there is, let's keep it like this

File: tests/device_tests/test_protect_call.py [hunk 23]
```diff
@@ -18,92 +18,53 @@
...
-
-        with pytest.raises(AssertionError), client:
-            # Scenario 3 - Not received expected message
-            client.set_expected_responses(
-                [proto.ButtonRequest(), proto.Success(), proto.Success()]
-            )  # This is expected, but not received
-            self._some_protected_call(client, True, False, False)
-
-        with pytest.raises(AssertionError), client:
+        with client:
```
Comment (matejcik) [inline tests/device_tests/test_protect_call.py]: if the `expected_responses` scenarios are removed, is there even a point in keeping this testcase?

File: legacy/firmware/config.c (hunk 22 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Indeed. At least that is what I am thinking. And also @prusnak mentioned it himself

File: tests/device_tests/test_protect_call.py (hunk 23 above)
Comment (tsusanka) [reply tests/device_tests/test_protect_call.py]: Hmm probably not. Removed

File: tests/device_tests/test_protection_levels.py [hunk 24]
```diff
@@ -261,23 +261,21 @@ def test_pin_cached(self, client):
         assert client.features.pin_cached is False
 
         with client:
-            client.set_expected_responses(
-                [proto.PinMatrixRequest(), proto.ButtonRequest(), proto.Address()]
-            )
-            btc.get_address(client, "Bitcoin", [], show_display=True)
+            client.set_expected_responses([proto.PinMatrixRequest(), proto.Address()])
+            btc.get_address(client, "Testnet", [0])
```
Comment (tsusanka) [inline tests/device_tests/test_protection_levels.py]: Just wondering - why did you remove Testnet above but left it here?

Comment (matejcik) [reply tests/device_tests/test_protection_levels.py]: basically because it's a required argument for `get_address`, see commit message for https://github.com/trezor/trezor-firmware/pull/803/commits/7377c204d61af74d2f3051972c272513a1911b37

and I wanted to make the diff the smallest, so I used the parameters that were most common in the original changeset, which is `"Testnet", [0]` as opposed to `[]` which is also allowed

**Comment by tsusanka** [comment]
I have replied to the comments, reviewed your commits and did a rebase.

File: core/src/apps/homescreen/__init__.py (hunk 21 above)
Comment (prusnak) [reply core/src/apps/homescreen/__init__.py]: Usability

File: legacy/firmware/config.c (hunk 22 above)
Comment (prusnak) [reply legacy/firmware/config.c]: Correct. AFAIK we don't support loading XPRV on TT, so let's drop this functionality from T1 as well.

Comment (matejcik) [reply legacy/firmware/config.c]: it's actually gone as of fd8aed610543d38f8f01091c2badd8af32c1d5f2, this code seems to be vestigial

Comment (tsusanka) [reply legacy/firmware/config.c]: Indeed. Marking as resolved.

File: legacy/firmware/protect.c (hunk 3 above)
Comment (prusnak) [reply legacy/firmware/protect.c]: No. It doesn't work because ppa->passphrase is of type `string`, not `bytes`. Strings in nanopb are zero-terminated `char *`, Bytes in nanopb are `struct { pb_size_t size; pb_byte_t bytes[n]; }`

File: core/embed/firmware/version.h [hunk 25]
```diff
@@ -4,6 +4,6 @@
 #define VERSION_BUILD 0
 
 #define FIX_VERSION_MAJOR 2
-#define FIX_VERSION_MINOR 3
+#define FIX_VERSION_MINOR 2
```
Comment (tsusanka) [inline core/embed/firmware/version.h]: How come? We have released 2.2.0 at the beginning of the year.

File: common/protob/messages-management.proto [hunk 26]
```diff
@@ -20,7 +20,7 @@ enum BackupType {
  * @next Features
  */
 message Initialize {
-    optional bytes session_id = 1;           // assumed device session id; Trezor clears caches if it is different or empty
+    optional bytes session_id = 3;           // assumed device session id; Trezor clears caches if it is different or empty
```
Comment (matejcik) [inline common/protob/messages-management.proto]: is this necessary? the `state` field that was 1 originally is functionally identical imo

File: core/tests/test_storage.cache.py [hunk 27]
```diff
@@ -82,14 +82,11 @@ def call_Initialize(**kwargs):
...
 
-        def call_ClearSession():
-            return await_result(handle_ClearSession(DUMMY_CONTEXT, ClearSession()))
-
         session_id = call_Initialize().session_id
         cache.set(KEY, "hello")
         self.assertEqual(cache.get(KEY), "hello")
 
-        call_ClearSession()
+        call_Initialize()
```
Comment (matejcik) [inline core/tests/test_storage.cache.py]: please drop the whole test case, it's not doing anything other than the `test_Initialize` above it

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 28]
```diff
@@ -107,8 +107,8 @@ def test_clear_session(client):
     assert session_id == new_session_id
 
     # by clearing the session, the id is lost
-    response = client.call(messages.ClearSession())
-    assert isinstance(response, messages.Success)
+    response = client.call(messages.Initialize())
+    assert isinstance(response, messages.Features)
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: again please drop the `clear_session` testcases completely, ISTM that all the cases are already covered in other tests

File: common/protob/messages-management.proto (hunk 26 above)
Comment (prusnak) [reply common/protob/messages-management.proto]: I agree, but this is cleaner and more explicit. If wallet pushed old field `1` Trezor will know it does not even have to check the field, because it comes from the old wallet. If there is a strong argument to keep `1` you'll be able to persuade me, though :-)

File: common/protob/messages-common.proto [hunk 29]
```diff
@@ -110,31 +111,16 @@ message PinMatrixAck {
...
-message PassphraseStateRequest {
-    optional bytes state = 1;       // actual device state
-}
-
-/**
- * Request: Send passphrase state back
- * @auxend
- */
-message PassphraseStateAck {
+    optional bool on_device = 2;    // user wants to enter passphrase on the device
```
Comment (matejcik) [inline common/protob/messages-common.proto]: the diff makes it difficult to see, but we changed a field type here.
orig: `optional bytes state = 2`
new: `optional bool on_device = 2`
it really should be 3

File: common/protob/messages-management.proto (hunk 26 above)
Comment (matejcik) [reply common/protob/messages-management.proto]: Here's an argument: if the id stays the same, I can write code that works with fw < 2.3.0 and fw >= 2.3.0 at the same time, without having a "if version < (2,3,0)" check

pre 2.3.0:
```
-> Initialize(session_id=ABCD)
<- Features(session_id=None)
-> GetAddress
<- PassphraseRequest(on_device=True)
-> PassphraseAck(on_device=True, passphrase=None, state=ABCD)
<- PassphraseStateRequest(state=EFGH)
-> PassphraseStateAck
<- Address
...
-> Initialize(session_id=EFGH)
```

2.3.0 and up:
```
-> Initialize(session_id=ABCD)
<- Features(session_id=EFGH)
-> GetAddress
<- PassphraseRequest
-> PassphraseAck(on_device=True, passphrase=None, state=EFGH)
<- Address
...
-> Initialize(session_id=EFGH)
```

I mean, to make it work I need to muck about with protobuf definitions locally anyway, so I can as well add back the `state = 1` and send both. But then the message will be twice as big.

If 2.3.0 is not a mandatory update, having such code in trezorlib would be useful, esp. for Electrum where it's difficult to tell people to "match your trezorlib version to your firmware version"

File: common/protob/messages-common.proto (hunk 29 above)
Comment (tsusanka) [reply common/protob/messages-common.proto]: Fixed in 56dc952495eac8ead7e750e60eded7096e41f6b2

File: common/protob/messages-management.proto (hunk 26 above)
Comment (prusnak) [reply common/protob/messages-management.proto]: What is the situation for T1? You'd need to have two different code paths for < 1.9.0 and >= 1.9.0, right?

Comment (matejcik) [reply common/protob/messages-management.proto]: T1 doesn't use any of these fields so I don't need to do anything special AFAICT. I decide whether to send `on_device=True` based on capabilities, which comes out right, and other than that the T1 flow is identical across versions.

Comment (matejcik) [reply common/protob/messages-management.proto]: speaking of, this might even be an argument for leaving the old definitions for fields and `PassphraseState` messages in the proto files (so that users don't need to hack the protobuf locally)
